src/test/java/teammates/test/cases/browsertests/StudentProfilePageUiTest.java

```diff
@@ -332,4 +333,26 @@ private StudentProfilePage getProfilePageForStudent(String studentId) {
         return loginAdminToPage(profileUrl, StudentProfilePage.class);
     }
 
+    @AfterClass
+    public void classTearDown() {
+        // The data associated with the test accounts have to be removed when the test ends because
+        // the test accounts are shared across tests. In particular, the courses of test student1 account is not the same as
+        // `StudentHomePageUiTest` so it has to be cleared before running that test.
+
+        // Test data is never cleared after test in order to save datastore costs as we already remove and restore the data
+        // bundle when the test runs. This raises the question of why we do not remove the data bundle when the class
+        // ends then we do not have to remove it from startup. The reason is that in the case a test fails to tear down
+        // (fail to remove data bundle), another test should have no reason to fail.
+
+        // This means that removing the data bundle on startup is not always sufficient because a test only knows how
+        // to remove its associated data. This is the reason why `StudentHomePageUiTest` would fail if we don't remove the
```

Comment:
I think a test should know how to create a 'clean slate' for itself at the start.

Reply:
Agreed, so it seems we have a constraint here.

- If it deletes too much data, it may affect other tests.
- If we delete more aggressively then we will probably lose some benefits of saving costs when deleting on startup.
- If we only delete more aggressively in the context of this test, then this test is treated as a special case.

If we expand the part on deleting more aggressively in the context of this test, we can generalize it to only when accounts are actual accounts we delete all their associated data. In that case, I think it's better but it seems like we are adding more magic to behind the scenes which hide the coupling of data.

On the other hand, I currently prefer @tshradheya's approach because it explicitly shows there's a coupling between tests and the data bundle has to be removed elsewhere because of the broken test design.

Although this won't be a problem if the data is not shared as you mentioned. For example, in the context of this test if we don't use shared accounts a test should be able to create a clean slate for itself.

---

Would you prefer instead of another test removing the data bundle that we add additional logic to delete all associated data if the accounts are actual accounts?

Reply:
How about we have a `deleteSharedTestData()` method that is called at the start of any test using shared data (i.e., test data related to actual accounts)? We have to ensure that such tests are run sequentially though.

Reply:
> How about we have a deleteSharedTestData() method that is called at the start of any test using shared data 

Some current tests actually do that. Like if some course is being used with same name in two places, in `processTestData()` there is a `Backdoor.deleteCourse()` but this isnt really ideal cause at times this will give a NullPOinterException cause maybe that course from another test data is not even created yet.

Reply:
I think `deleteSharedTestData()` is something possible to consider (maybe can do some checking) but we don't have to do it in this PR because we would have to address the other tests doing the same thing and we are not be solving some of the related issues. Let's keep this in view for the follow-up issue?